rfc/002-fault-tolerance/README.md

```diff
@@ -0,0 +1,487 @@
+operation may remove N partitions and replace them with a new one. These
+operations break down to a sequence of deletes, followed by an insert.
+
+### Non-requirements
+
+Infrastructure approaches for hardware redundancy are considered out of scope.
+In particular, we do not consider fault tolerance at the storage layer. There
+are many different ways to provide redundant storage, e.g., RAID storage,
+distributed file systems or NAS. Cloud storage, like S3 or Azure blob storage,
+are replicated three times, and adding an additional layer of fault tolerance on
+top does not provide a lot of additional safety but considerably increases the
+cost.
+
+## Solution Proposal
+
+The most important architectural effort concerns making the catalog a
+fault-tolerant component. To this end, we propose to factor the logic that
+operates on partition metadata into a dedicated **metastore** component. As a
+first point of contact for queries, the catalog also keeps per-partition index
+structures in memory. This catalog functionality is out of scope, as it
+represents derived, immutable state from the metastore.
+
+Adjacent fault-tolerance efforts outside the catalog scope involve replication
+of components. For example, the failure domain of the read path is a single
+query, which we can deploy in a dedicated operating system process (that extends
+to a single cloud function or container execution). Likewise, we can isolate
+failure domains on the write path by spawning each data source (or pipeline) in
+its own process. This flexibility is a by-product of the actor model
+architecture that treats internal components as microservices. The spectrum of
+fate sharing is the OS process boundary. The choice highly depends on the
+specifics of the deployment environment.
+
+We briefly touch on how the mechanism for making the write and read path
+resilient, and then focus the discussion on fault tolerance within the VAST
+server node, specifically the catalog.
+
+### Resilient Write Path
+
+Data sources provide a stream of events. VAST parses them and bundles them into
+units of partitions. Assuming that we can build a fault-tolerant catalog, the
+problem of making the write path fault-tolerant reduces to ensuring exactly-once
+processing semantics on the ingest stream.
+
+If the circumstances require high-fidelity data acquisition, we can resort to an
+external data plane with exactly-once processing stream semantics, such as
+Apache Kafka. VAST must not acknowledge events from the data plane until having
+received a confirmation that the corresponding partition is durable. In the
+presence of a crash, VAST must restart from the last acknowledged position in
+the stream.
+
+Because security telemetry data is already lossy upstream (e.g., due to packet
+loss or lacking instrumentation), operators are often willing to accept looser
+guarantees. Therefore, we deem it sufficient to offer a mechanism that
+implements interaction with a durable message bus.
```

Comment:
As @rdettai already wrote [in another comment](https://github.com/tenzir/vast/pull/2562/files#r1003186045), this isn't actually the correct way to establish exactly once. I'll change this paragraph accordingly.

I also don't want to go into the various options in detail here, because there are tradeoffs to be explained for each possible level of granularity and the additional text would only distract from the center of focus.

In the end, the entire section here is about dividing the problem into sub-problems. In the end, this RFC mainly concerns itself with the ground truth of metadata storage. If we need further architectural changes to harden the write path they should best be discussed in a separate RFC.

Reply:
TODO: Replace the last sentence with::
```
This may require a unique identifier for every data source, possibly requiring the user to provide one explicitly.
```